lens/fetcher.go

```diff
@@ -0,0 +1,397 @@
+			if !ok {
+				return core.Doc{}, core.ErrInvalidKey
+			}
+
+			doc.SetKey(key)
+			continue
+		}
+
+		fieldDesc, fieldFound := df.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.Decode(fieldDesc, fieldByteValue)
+		if err != nil {
+			return core.Doc{}, err
+		}
+
+		index := mapping.FirstIndexOfName(fieldName)
+		doc.Fields[index] = fieldValue
+	}
+
+	if value, ok := docAsMap[request.DeletedFieldName]; ok {
+		if wasDeleted, ok := value.(bool); ok {
+			if wasDeleted {
+				doc.Status = client.Deleted
+			} else {
+				doc.Status = client.Active
+			}
+		}
+	}
+
+	doc.SchemaVersionID = df.col.Schema.VersionID
+
+	return doc, nil
+}
+
+// lensDocToClientDoc converts a LensDoc to a client.Document.
+func (df *lensedFetcher) lensDocToClientDoc(docAsMap LensDoc) (*client.Document, error) {
+	key, err := client.NewDocKeyFromString(docAsMap[request.KeyFieldName].(string))
+	if err != nil {
+		return nil, err
+	}
+	doc := client.NewDocWithKey(key)
+
+	for fieldName, fieldByteValue := range docAsMap {
+		if fieldName == request.KeyFieldName {
+			continue
+		}
+
+		fieldDesc, fieldFound := df.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.Decode(fieldDesc, fieldByteValue)
+		if err != nil {
+			return nil, err
+		}
+
+		err = doc.SetAs(fieldDesc.Name, fieldValue, fieldDesc.Typ)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	doc.SchemaVersionID = df.col.Schema.VersionID
+
+	// Note: client.Document does not have a means of flagging as to whether it is
+	// deleted or not, and, currently the fetcher does not ever returned deleted items
+	// from the function that returs this type.
+
+	return doc, nil
+}
+
+// updateDataStore updates the datastore with the migrated values.
+//
+// This removes the need to migrate a document everytime it is fetched as the second time around
+// the underlying fetcher will return the migrated values cached in the datastore instead of the
+// underlying dag store values.
+func (df *lensedFetcher) updateDataStore(ctx context.Context, original map[string]any, migrated map[string]any) error {
+	modifiedFieldValuesByName := map[string]any{}
+	for name, originalValue := range original {
+		migratedValue, ok := migrated[name]
+		if !ok {
+			// If the field is present in the original, and missing from the migrated, it
+			// means that a migration has removed it, and we should set it to nil.
+			modifiedFieldValuesByName[name] = nil
+			continue
+		}
+
+		// Note: A deep equals check is required here, as the values may be inline-array slices
+		if !reflect.DeepEqual(originalValue, migratedValue) {
```

Comment:
suggestion: very hesitant to use `reflect.DeepEqual` as its absurdly ineffecient. I understand the want to use it, but considering this is a likely branch in a hot path, it would be nice to avoid. Obviously its not as computationally expensive as the actual lens migration, but thats just a better reason to avoid additional unnecessary overhead.

An alternative approach would certainly require a decent bit of code, theres likely some generics sugar that can simplify it with the `comparable` type constraint, as there as only a handful of values/types the doc fields can use.

Reply:
I really didnt want to use this either, and had a comment noting that, which I appeared to have trimmed. Are you happy with a code-comment todo linked to a ticket, or do you want this changed before merge?

- [ ] At least document how crap an undesireable reflect.DeepEquals is

Reply:
There is enough in this PR, and this is really not a critical issue that should block the release of Lens.

Reply:
I counted out of curiousity, there are 15 total in this branch, of which 6 after for the same issue, and 5 are introduced by this PR.  Meaning that in develop branch there are about 5 outstanding unique todo comments.

Reply:
not to be pedantic, but there are lots of "followup" issues that aren't specifically linked to a `todo` comment in the code, mentioned this in a general way the other day in DMs with you, hesitant to continue to this.

I can tolerate delaying this, but do please keep it on your mind for (high priority) follow up work. In a perfect world I would request to do this now in PR, cus I truely despise this function with a burning passion, and only really tolerate in test assertions, let alone in a likely hot path for query execution.

Reply:
> not to be pedantic, but there are lots of "followup" issues that aren't specifically linked to a todo comment in the code, mentioned this in a general way the other day in DMs with you, hesitant to continue to this.

This was an earlier problem, and since as a team we agreed to never commit todos unlinked to an issue I think the number has been sensible, and it has been a good way to preserve PR focus and note flaws or bugs that can be solved later.  I have close quite a few of these linked todos.  The unlinked ones I would happily purge from the codebase.

Reply:
Does this need an explicit `todo` issue?

Reply:
It has one, linked in the PR description and via a code comment.